src/file-manager/fm-desktop-wayland-bg-dialog.c

```diff
@@ -0,0 +1,418 @@
+
+    if (!provider)
+        provider =  gtk_css_provider_new ();
+
+    /*Build a color preview using a cssprovider due to requirement to handle RBGA values*/
+    string = g_string_new(NULL);
+    g_string_append (string, "#caja-wayland-bg-preview {");
+
+    if (strcmp (shading_type, "vertical-gradient") == 0)
+        g_string_append (string, "background-image: linear-gradient(to bottom,");
+
+    else if  (strcmp (shading_type, "horizontal-gradient") == 0)
+        g_string_append (string, "background-image: linear-gradient(to right,");
+
+    else
+    {
+        g_string_append (string, "background-color:");
+        g_string_append (string, primary_color_str);
+    }
+
+    if ((strcmp (shading_type, "vertical-gradient") == 0) ||
+       (strcmp (shading_type, "horizontal-gradient") == 0))
+    {
+        g_string_append (string, primary_color_str);
+        g_string_append (string, ",");
+        g_string_append (string, secondary_color_str);
+        g_string_append (string, ");");
+    }
+        g_string_append (string, "}");
+
+    css = g_string_free (string, FALSE);
+    gtk_style_context_add_provider (gtk_widget_get_style_context (preview),
+                                    GTK_STYLE_PROVIDER (provider),
+                                    GTK_STYLE_PROVIDER_PRIORITY_APPLICATION);
+
+    gtk_css_provider_load_from_data (provider, css, -1, NULL);
+
+    if ((!filename) || (strcmp (filename, "") == 0) || (strcmp (filename, " ") == 0))
+        gtk_image_clear(GTK_IMAGE(preview_image));
+
+    else
+    {
+        GdkRectangle geometry = {0};
+        GdkPixbuf  *pixbuf;
+        GdkMonitor *monitor;
+        GdkDisplay *display;
+
+        display = gdk_screen_get_display (gdk_screen_get_default());
+        monitor = gdk_display_get_monitor (display, 0);
+        gdk_monitor_get_geometry (monitor, &geometry);
+
+        gtk_style_context_remove_provider (gtk_widget_get_style_context (preview),
+                                           GTK_STYLE_PROVIDER (provider));
+
+        gtk_widget_set_size_request (preview, geometry.width / 5, geometry.height / 5);
+
+        pixbuf = gdk_pixbuf_new_from_file_at_scale (filename, geometry.width / 5, 
+                                                    geometry.height / 5, FALSE, NULL);
+
+        gtk_image_set_from_pixbuf (GTK_IMAGE (preview_image), pixbuf);
+        g_object_unref(pixbuf);
+    }
+}
+
+static void
+update_primary_color(GtkWidget *colorbutton1)
+{
+    GdkRGBA color1;
+
+    filename = "";
+    gtk_color_chooser_get_rgba (GTK_COLOR_CHOOSER (colorbutton1), &color1);
+    primary_color_str = gdk_rgba_to_string (&color1);
+    g_settings_set_string (background_settings,
+                           "primary-color", primary_color_str);
+
+    g_settings_set_string (background_settings,
+                       "picture-filename", "");
+    update_preview();
+}
+
+static void
+update_secondary_color(GtkWidget *colorbutton2)
+{
+    GdkRGBA color2;
```

Comment:
Build warnings in CI:
https://app.travis-ci.com/github/mate-desktop/caja/jobs/615896488#L6018
```
Checking src/file-manager/fm-desktop-wayland-bg-dialog.c: HAVE_WAYLAND...

src/file-manager/fm-desktop-wayland-bg-dialog.c:135:13: style: Local variable 'color1' shadows outer variable [shadowVariable]

    GdkRGBA color1;
            ^

src/file-manager/fm-desktop-wayland-bg-dialog.c:59:14: note: Shadowed declaration

GdkRGBA      color1, color2;
             ^

src/file-manager/fm-desktop-wayland-bg-dialog.c:135:13: note: Shadow variable

    GdkRGBA color1;
            ^

src/file-manager/fm-desktop-wayland-bg-dialog.c:151:13: style: Local variable 'color2' shadows outer variable [shadowVariable]

    GdkRGBA color2;
            ^

src/file-manager/fm-desktop-wayland-bg-dialog.c:59:22: note: Shadowed declaration

GdkRGBA      color1, color2;
                     ^

src/file-manager/fm-desktop-wayland-bg-dialog.c:151:13: note: Shadow variable

    GdkRGBA color2;
```

Reply:
Both of those local declarations have been removed. I had to make these globals to be able to access them from multiple functions, due to limitations in how many variables could be easily passed with` g_signal_connect`